theano/scan_module/scan_op.py

```diff
@@ -63,6 +63,7 @@ def __init__(self,
         # adding properties into self
         self.inputs = inputs
         self.outputs = outputs
+        self.output_is_input = [out in inputs for out in outputs]
```

Comment:
I think this isn't the complete fix.

I think I understand more the problems. Around line 910-932, we put in the input_storage some value that come from outs. It happen in the test case from test_borrow_bug_jeremiah, that this data set in input_storage is also set in output_storage. So we override the input. If this work or not depend on the order of the node in the graph and in that case, the order is random. So we are lucky that it make this test fail !

I think we must analyze more when we put in the input_storage something from outs and more sure that we do not put in the output_storage what is currently in the input_storage or will be in the intput_storage at a later step. We could in theory allow some input_storage to point to the same data as the output_storage in some case. But 1) I think it is hard to do right and 2) I think it won't be used frequently. So I think we do not need to optimize now this case.

@carriepl, I let you look more in detail on that and try to find the additional condition we need to do.